be/src/fs/s3/poco_common.cpp

```diff
@@ -97,4 +77,63 @@ bool checkRequestCanReturn2xxAndErrorInBody(Aws::Http::HttpRequest& request) {
     return false;
 }
 
+HTTPSessionPtr makeHTTPSessionImpl(const std::string& host, Poco::UInt16 port, bool https, bool keep_alive) {
+    HTTPSessionPtr session;
+
+    if (https) {
+        session = std::make_shared<Poco::Net::HTTPSClientSession>(host, port);
+    } else {
+        session = std::make_shared<Poco::Net::HTTPClientSession>(host, port);
+    }
+
+    // doesn't work properly without patch
+    session->setKeepAlive(keep_alive);
+    return session;
+}
+
+EndpointHTTPSessionPool::Base::ObjectPtr EndpointHTTPSessionPool::allocObject() {
+    SCOPED_THREAD_LOCAL_MEM_TRACKER_SETTER(_mem_tracker);
+    auto session = makeHTTPSessionImpl(_host, _port, _is_https, true);
+    return session;
+}
+
+HTTPSessionPools& HTTPSessionPools::instance() {
+    static HTTPSessionPools instance;
+    return instance;
+}
+
+PooledHTTPSessionPtr HTTPSessionPools::getSession(const Poco::URI& uri, const ConnectionTimeouts& timeouts,
+                                                  bool resolve_host) {
+    const std::string& host = uri.getHost();
+    uint16_t port = uri.getPort();
+    bool is_https = isHTTPS(uri);
+
+    const Key key = {.host = host, .port = port, .is_https = is_https};
+
+    EndpointPoolPtr pool = nullptr;
+    {
+        std::lock_guard lock(_mutex);
```

Comment:
We have many IO threads. Will this lock introduce heavy lock contention?

Reply:
I will do a test on a multi-core instance.

Reply:
Have done a test on a 48 cores instance, the lock contention is not serious, and it's faster than default curl client while use less cpu.